python/lsst/daf/butler/registry/interfaces/_database.py

```diff
@@ -1430,10 +1432,68 @@ def delete(self, table: sqlalchemy.schema.Table, columns: Iterable[str], *rows:
             # while reporting that no rows were affected.
             return 0
         sql = table.delete()
-        whereTerms = [table.columns[name] == sqlalchemy.sql.bindparam(name) for name in columns]
-        if whereTerms:
-            sql = sql.where(sqlalchemy.sql.and_(*whereTerms))
-        return self._connection.execute(sql, *rows).rowcount
+        columns = list(columns)  # Force iterators to list
+
+        # More efficient to use IN operator if there is only one
+        # variable changing across all rows.
+        content: Dict[str, Set] = defaultdict(set)
+        if len(columns) == 1:
+            # Nothing to calculate since we can always use IN
+            column = columns[0]
+            changing_columns = [column]
+            content[column] = set([row[column] for row in rows])
+        else:
+            for row in rows:
+                for k, v in row.items():
+                    content[k].add(v)
+            changing_columns = [col for col in content if len(content[col]) > 1]
+
+        if len(changing_columns) == 1:
+            name = changing_columns.pop()
+
+            # Simple where clause for the unchanging columns
+            clauses = []
+            for k, v in content.items():
+                if k == name:
+                    continue
+                column = table.columns[k]
+                # The set only has one element
+                clauses.append(column == v.pop())
+
+            # The IN operator will not work for "infinite" numbers of
+            # rows so must batch it up into distinct calls.
+            in_content = list(content[name])
+            n_elements = len(in_content)
+
+            # This is a soft limit which might be increased by 50%.
+            # For example with max=1001 if 1500 elements are involved
+            # the code will choose one loop of 1500. If 2000 elements are
+            # involved there will be two loops of 1000.
+            # The idea is to allow some adjustment of the limits so that
+            # if 1002 elements are present this is done in an even split
+            # and not as 1001 elements in one call and 1 in another.
+            #
+            MAX_ELEMENTS_PER_IN = 1000
+            n_loops = round(n_elements / MAX_ELEMENTS_PER_IN)
+            if n_loops == 0:
+                n_loops = 1
+            n_per_loop = math.ceil(n_elements / n_loops)
+
+            rowcount = 0
+            iposn = 0
+            while iposn < n_elements:
+                endpos = iposn + n_per_loop
+                in_clause = table.columns[name].in_(in_content[iposn:endpos])
+                iposn = endpos
```

Comment:
It maybe easier to read as:

    for iposn in range(0, n_elements, n_per_loop):
        endpos = iposn + n_per_loop